### DISCUSSION TITLE
BullMQ worker not always triggered with Redis hosted on VercelKV

### DISCUSSION THREAD

**Discussion by adrianfabjanski** [vercel/community#5050] [Help] [answered]
### Summary

I'm quite new to job scheduling and delayed operations. I have a next.js app running on vercel.

I have a messenger built into my app, so when the message is sent, I was sending email informing about a new message inside the same request in route.ts in NextJS.

Obviously it was not the best idea.

First, sending a message was taking a lot of time. Second, I don't want to send an email for every message. I wanted to create a "time window", in which user can send messages. Every time he sends another message, the delay is being reset. And it seemed like working..
But, I placed a couple of console.logs in my queue file informing the current state of the job. For one message, there was a job added to the queue, but the worker never got triggered.

Sometimes worker is being triggered, but in 30 minutes, instead of 1.5 minutes.

So what I did is I created a VercelKV redis storage, and installed bullMQ on my next app. Then Im creating a new Queue with ioredis, and passing the redis connection string from Vercel. Then when sending a message, Im adding a new job to the queue with the id being the recipients email.

If there is already a job with this email (when user is sending another message. The "time window" I mentioned), and the job is delayed, I'm resetting the delay to the original one, which in this case is 9000ms (1.5 minutes).

when the job is completed or failed, im removing the job so when the user sends a new message for the same email, there is no job with this id.

What am I missing? The problem is not consistent, sometimes it sends 1.5 minutes later spot on. Sometimes 30 minutes. Sometimes never.

Today I had this issue that a message was sent in the morning, job was created, but the worker didn't trigger. Another message was sent in a different chat later, and that triggered the worker. 
It looked like the worker got triggered when a new connection was established. 

### Steps to Reproduce

This is my code for emailqueue.ts. Im using it in my send_message next.js route with await addEmail({..})

```
import { Job, Queue, Worker } from "bullmq";
import IORedis from "ioredis";
import { Resend } from "resend";
import { EmailType, resolveTemplate } from "./resolveTemplate";

const resend = new Resend(process.env.RESEND_API_KEY);

const JOB_DELAY = 90000;

export interface QueueEmail {
  to: string;
  subject: string;
  componentProps: any;
  emailType: EmailType;
}

const messageQueue = new Queue("messages", {
  connection: new IORedis(process.env.REDIS_CONNECTION_STRING!, {
    maxRetriesPerRequest: null,
    enableReadyCheck: false,
  }),
});

messageQueue.on("error", (error) => {
  console.error(`Queue error: ${error}`);
});

messageQueue.on("waiting", (job) => {
  console.log(`Job ${job.id} added`);
});

// This function adds a message to the queue
export const addEmail = async (message: QueueEmail) => {
  const emailTo = message.to as string;
  const existingJob = await messageQueue.getJob(emailTo);
  if (existingJob) {
    console.log(`job for ${existingJob.id} exists`);
    if (await existingJob.isDelayed()) {
      console.log(`delaying ${existingJob.id} for another ${JOB_DELAY}ms`);
      existingJob.changeDelay(JOB_DELAY);
      console.log(`current delay: ${existingJob.delay}ms`);
    } else {
      const jobState = await existingJob.getState();
      console.log(`current job for ${existingJob.id}: ${jobState}`);
    }
  } else {
    await messageQueue.add(
      "sendEmail",
      { message },
      { delay: JOB_DELAY, jobId: emailTo }
    );
    console.log(`Job added to the queue for ${emailTo}`);
  }
};

// Create a new worker to process jobs from the queue
const worker = new Worker(
  "messages",
  async (job: Job) => {
    const { message } = job.data as { message: QueueEmail };
    const emailTo = message.to as string;

    console.log(`Sending email to ${emailTo}`);
    resend.emails.send({
      from: "someemail.com <noreply@someemail.com>",
      to: emailTo,
      subject: message.subject,
      react: resolveTemplate(message.emailType, message.componentProps),
    });
  },
  {
    connection: new IORedis(process.env.REDIS_CONNECTION_STRING!, {
      maxRetriesPerRequest: null,
      enableReadyCheck: false,
    }),
  }
);

const removeJob = async (job?: Job) => {
  console.log(`Removing job for ${job?.id}..`);
  await job?.remove();
  console.log("Job removed!");
};

worker.on("completed", async (job) => {
  console.log(`Job completed for ${job.id}`);
  await removeJob(job);
});
worker.on("failed", async (job, err) => {
  console.error(`${job?.id} has failed with ${err.message}`);
  a
```

**Comment by adrianfabjanski** [comment, marked as answer]
Turns out you have to use the `instrumentationHook` in order to make this work easily. 
Solution here: https://medium.com/@mofaddelzakaria/nextjs-with-bullmq-fc25aa7b8fe5

**Reply by yovanoc** [reply]
link deleted, I have this problem too, what is instrumentationHook, don't found it on bullmq docs

**Reply by adrianfabjanski** [reply]
It's an experimental feature of next.js:
https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation

You have to put your worker inside  **instrumentation.ts** file that is in your root directory.

For me it looks like this:

```
//instrumentation.ts
export const register = async () => {
  if (process.env.NEXT_RUNTIME === "nodejs") {
    const { Worker } = await import("bullmq");
    const { connection } = await import("./redis/connection");
    const { emailWorker } = await import("./redis/email_worker");

    // Create a new worker to process jobs from the queue
    const worker = new Worker("messages", emailWorker, { connection });

    worker.on("completed", async (job) => {
      console.log(`Job completed for ${job.id}`);
    });
    worker.on("failed", async (job, err) => {
      console.error(`${job?.id} has failed with ${err.message}`);
    });
    worker.on("stalled", (str) => {
      console.log(`Job stalled: ${str}`);
    });
  }
};
```

**Reply by masterjanic** [reply]
Thanks, this also helped me!

**Reply by Adel-Alwaqfi** [reply]
@adrianfabjanski 
Hi, could you please share the code of the emailWorker you're using?
I have the same problem and I need to know how this part constructed

**Reply by adrianfabjanski** [reply]
Hi,

The code I was using looked like this:

```
// email_worker.ts
import sendEmail from "@/app/api/_shared/email_service";
import { render } from "@react-email/render";
import { Job } from "bullmq";
import { QueueEmail } from "./enqueue_email";
import { resolveTemplate } from "./resolve_template";

export const emailWorker = async (job: Job) => {
  const { message } = job.data as { message: QueueEmail };
  const emailTo = message.to as string;

  console.log(`Sending email to ${emailTo}`);
  await sendEmail({
    to: emailTo,
    subject: message.subject,
    html: render(resolveTemplate(message.emailType, message.componentProps)),
  });
};

// connection.ts
import IORedis from "ioredis";

export const connection = new IORedis(process.env.REDIS_CONNECTION_STRING!, {
    maxRetriesPerRequest: null,
    enableReadyCheck: false,
  });

// email_queue.ts
import { Queue } from "bullmq";
import { connection } from './connection';

export const messageQueue = new Queue("messages", {
    connection,
  });
  
  messageQueue.on("error", (error) => {
    console.error(`Queue error: ${error}`);
  });
  
  messageQueue.on("waiting", (job) => {
    console.log(`Job ${job.id} added`);
  });

```

**Comment by pushkarsingh32** [comment]
hi the worker handle is not able to detect localhost api any idea? 

```
export const articleWorkerHandler = async (job) => {
    let { articleId, allParams, } = job.data;
    console.info(`I am processing the job with id ${job.id} & articleId ${articleId}`);
    const siteHomepageUrl = process.env.NEXT_PUBLIC_SITE_HOMEPAGE_URL;

    let writerAPILocation = siteHomepageUrl + '/api/writer';
    console.log(`Sending request to writer API at ${writerAPILocation}`);
    console.info(`all params are: ${JSON.stringify(allParams)}`);
  
    await sendBulkWriterRequest(allParams, writerAPILocation, articleId);

    return job.data;
};

```

**Reply by adrianfabjanski** [reply]
**I think** if your redis instance is not running on your local machine, but somewhere in the cloud (like in my case vercel), the worker won't be able to access the localhost. 

You would have to expose your localhost to the internet. 
You can do so by using **ngrok** for example. And then pass the generated url to the worker.
